contributors/design-proposals/admission_control_extension.md

```diff
@@ -0,0 +1,265 @@
+
+
+### Specification
+
+Based on observation of the actual admission control implementations the majority of mutation
+occurs as part of creation, and a large chunk of the remaining controllers are for side-effect free
+validation of creation and updates. Therefore we propose the following changes to Kubernetes:
+
+1.  Allow some controllers to act as "initializers" - watching the API and mutating the object before it is visible to normal clients.
+    This would reuse the majority of the infrastructure in place for controllers. Because creation is
+    one-way, the object can be "revealed" to regular clients once a set list of initializers is consumed. These
+    controllers could run on the cluster as pods.
+2.  Add a generic **external admission webhook** controller that is non-mutating (thus parallelizable)
+    This generic webhook API would resemble `admission.Interface` and be given the input object (for create) and the
+    previous object (for update). After initialization or on any update, these hooks would be invoked in parallel
+    against the remote servers and any rejection would reject the mutation.
+3.  Make the registration of both initializers and admission webhooks dynamic via the API (a configmap or cluster scoped resource)
+    Administrators should be able to dynamically add or remove hooks and initializers on demand to the cluster.
+    Configuration would be similar to registering new API group versions and include config like "fail open" or
+    "fail closed".
+
+Some admission controller types would not be possible for these extensions:
+
+* Mutating admission webhooks could be a later addition
+* Admission controllers that need access to the acting user can receive that via the external webhook.
+* Admission controllers that "react" to the acting user can couple the information received via a webhook and then act if they observe mutation succeed (tuple combining resource UID and resource generation).
+* Quota will continue to be a core plugin per API server, so extension is not critical.
+
+
+
+#### Implications:
+
+1. Initializers and generic admission controllers are highly privileged, so while some separation is valuable they are effectively cluster scoped
+2. This mechanism would allow dedicated infrastructure to host admission for multiple clusters, and allow some expensive admission to be centralized (like quota which is hard to performantly distribute)
+3. There is no way to build initializers for updates without a much more complicated model, but we anticipate initializers to work best on creation.
+4. Ordering will probably be necessary on initializers because defaulting in the wild requires ordering. Non-mutating validation on the other hand can be fully parallel.
+5. Some admission depends on knowing the identity of the actor - we will likely need to include the **creator** as information to initializers.
```

Comment:
I think field level authorization is an admission controller that is earlier in the chain and not necessarily part of this problem statement.  It is not necessary to invert the core api server admission to allow a third party to impose field level authorization if we design field level authorization correctly.  All other specialized use cases are handled by external ordering.

In general, anything like field level authorization would be a non-mutating admission controller, so those would not be granted the rights to modify anything.  

Initializer A in your example has to go through normal admission.  So it can't escape anything that a normal update would allow.